File: RPi_Scala_Java_workspace/RPi_Scala_Java_project/src/main/scala/apps/BlinkLEDs.scala
package apps

/* Expand this comment for suggested program CLI argument and MV arguments
 * 
 * CLI and VM arguments can be supplied to eclipse urns by first establishing a "run configuration"
 * by clicking on the icon bar Run icon pulldown... run as...   scala application
 * 
 * THereafter icon bar Run iconpulldown... run configuratons... arguments tab
 * shows two sections:   the upper for CLI and the lower for VM
 * In the upper (program argument:) box is for CLI arguments
 * 
 * The lower (VM:) entryField has arguments fo the JVM
 * These come into play frequently in spark applications but are not expected to have
 * many applications on the RaspberryPi
 */

import org.slf4j.Logger._
import org.slf4j.LoggerFactory._
import org.slf4j.LoggerFactory
import java.time.format._
import java.time.LocalDateTime

import com.pi4j.system._
import com.pi4j.util._
import com.pi4j.util.Console

import com.pi4j.io.gpio.GpioController
import com.pi4j.io.gpio.GpioFactory
import com.pi4j.io.gpio.GpioPinDigitalInput
import com.pi4j.io.gpio.GpioPinDigitalOutput
import com.pi4j.io.gpio.PinPullResistance
import com.pi4j.io.gpio.PinState
import com.pi4j.io.gpio.RaspiPin
import com.pi4j.io.gpio.event.GpioPinDigitalStateChangeEvent
import com.pi4j.io.gpio.event.GpioPinListenerDigital


object BlinkLEDs {
	val appStart=java.time.LocalDateTime.now.format(DateTimeFormatter.ofPattern("yyyyMMdd_HH_mm_ss"))
	println("appStart at "+appStart)
	System.setProperty("appStart",appStart)
	/* these two variables are used in the log4j.properties file (defaults as /src/main/resource/log4j.properties when run from eclipse) */
	System.setProperty("invariantLogFileName","myAppLog")
	System.setProperty("uniquePerRunLogFileName","myAppLog_"+appStart)
	@transient lazy val log = org.slf4j.LoggerFactory.getLogger(this.getClass.getName.substring(0,this.getClass.getName.length-1))
	
	def main(args: Array[String]) = {
		if(0<args.length)for((a,ii)<-args.zipWithIndex)(println(f"args[$ii%2d]=$a%s")) else println("no CLI args")
		
		println("<--Pi4J--> GPIO Blink Example ... started .")
		
		// create gpio controller
		val gpio = GpioFactory.getInstance()
		
		// provision gpio pin #01 & #03 as an output pins and blink
		val led0 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_29) /* red */
		val led1 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_28) /* green */
		val led2 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_27) /* yellow */
		val led3 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_26) /* IR */

		// provision gpio pin #02 as an input pin with its internal pull down resistor enabled
		val myButton = gpio.provisionDigitalInputPin(RaspiPin.GPIO_02, PinPullResistance.PULL_DOWN)

		// create and register gpio pin listener
		myButton.addListener(new GpioPinListenerDigital() {
			override def handleGpioPinDigitalStateChangeEvent(event : GpioPinDigitalStateChangeEvent) {
				// when button is pressed, speed up the blink rate on LED #2
				if(event.getState().isHigh()){
					led2.blink(200)
				} else {
					led2.blink(1000)
				}
			}
		})

		// continuously blink the led every 1/4 second for 15 seconds: it's too bright to leave blinking 
		led0.blink(250, 15000)
		// continuously blink the led s 
		//led0.blink(250)
		led1.blink(500)
		led2.blink(1000)
		led3.blink(2000)
		
		System.out.println(" ... the LEDs will continue blinking until the program is terminated.")
		System.out.println(" ... PRESS <CTRL-C> TO STOP THE PROGRAM.")
				
		// keep program running until user aborts (CTRL-C)
		while(true) {
				Thread.sleep(250)
		}
		
//		/* was not able to achieve reproducable results with the setShutdownOptions */
//		// set shutdown state for these pins
//		led0.setShutdownOptions(false,PinState.HIGH)  
//		led1.setShutdownOptions(false,PinState.HIGH)
//		led2.setShutdownOptions(false,PinState.HIGH)
//		led3.setShutdownOptions(false,PinState.HIGH)
		
		// stop all GPIO activity/threads
		// (this method will forcefully shutdown all GPIO monitoring threads and scheduled tasks)
		//gpio.shutdown  /*  implement this method call if you wish to terminate the Pi4J GPIO controller */
		
				

	}
}

File: RPi_Scala_Java_workspace/RPi_Scala_Java_project/src/main/java/myJavaStuff/BlinkLEDs.java
package myJavaStuff;

//START SNIPPET: blink-gpio-snippet


/*
* #%L
* **********************************************************************
* ORGANIZATION  :  Pi4J
* PROJECT       :  Pi4J :: Java Examples
* FILENAME      :  BlinkGpioExample.java
*
* This file is part of the Pi4J project. More information about
* this project can be found here:  http://www.pi4j.com/
* **********************************************************************
* %%
* Copyright (C) 2012 - 2016 Pi4J
* %%
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Lesser Public License for more details.
*
* You should have received a copy of the GNU General Lesser Public
* License along with this program.  If not, see
* <http://www.gnu.org/licenses/lgpl-3.0.html>.
* #L%
*/


import com.pi4j.io.gpio.GpioController;
import com.pi4j.io.gpio.GpioFactory;
import com.pi4j.io.gpio.GpioPinDigitalInput;
import com.pi4j.io.gpio.GpioPinDigitalOutput;
import com.pi4j.io.gpio.PinPullResistance;
import com.pi4j.io.gpio.PinState;
import com.pi4j.io.gpio.RaspiPin;
import com.pi4j.io.gpio.event.GpioPinDigitalStateChangeEvent;
import com.pi4j.io.gpio.event.GpioPinListenerDigital;
/**
 * This example code demonstrates how to perform simple
 * blinking LED logic of a GPIO pin on the Raspberry Pi
 * using the Pi4J library.
 *
 * @author Robert Savage
 * modified slightly by TPMoyer to change to 4 LED's on the highest  GPIO numbered pins
 */
public class BlinkLEDs {

	public static void main(String[] args) throws InterruptedException {

        System.out.println("<--Pi4J--> GPIO Blink Example ... started .");

        // create gpio controller
        final GpioController gpio = GpioFactory.getInstance();

        // provision gpio pin #01 & #03 as an output pins and blink
        final GpioPinDigitalOutput led0 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_29); /* red */
        final GpioPinDigitalOutput led1 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_28); /* green */
        final GpioPinDigitalOutput led2 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_27); /* yellow */
        final GpioPinDigitalOutput led3 = gpio.provisionDigitalOutputPin(RaspiPin.GPIO_26); /* IR */

        // provision gpio pin #02 as an input pin with its internal pull down resistor enabled
        final GpioPinDigitalInput myButton = gpio.provisionDigitalInputPin(RaspiPin.GPIO_02, PinPullResistance.PULL_DOWN);

        // create and register gpio pin listener
            myButton.addListener(new GpioPinListenerDigital() {
                    @Override
                    public void handleGpioPinDigitalStateChangeEvent(GpioPinDigitalStateChangeEvent event) {
                        // when button is pressed, speed up the blink rate on LED #2
                        if(event.getState().isHigh()){
                          led2.blink(200);
                        }
                        else{
                          led2.blink(1000);
                        }
                    }
                });
       
//        // set shutdown state for this pin
//        PinState offState=PinState.HIGH; 
//        led0.setShutdownOptions(true, offState);
//        led1.setShutdownOptions(true, offState);
//        led2.setShutdownOptions(true, offState);
//        led3.setShutdownOptions(true, offState);
        
        
         //blink the led every 1/4 second for 15 seconds and then turn off (the red became distracting as it was so bright)
         led0.blink(250, 15000);

        // continuously blink the leds 500 miliseconds, 1 second, and 2 seconds
        //led0.blink(250);
        led1.blink(500);
        led2.blink(1000);
        led3.blink(2000);

        System.out.println(" ... the LEDs will continue blinking until the program is terminated.");
        System.out.println(" ... PRESS <CTRL-C> TO STOP THE PROGRAM.");

        // keep program running until user aborts (CTRL-C)
        while(true) {
            Thread.sleep(250);
        }

        // stop all GPIO activity/threads
        // (this method will forcefully shutdown all GPIO monitoring threads and scheduled tasks)
        // gpio.shutdown();   <--- implement this method call if you wish to terminate the Pi4J GPIO controller
    }
}

//END SNIPPET: blink-gpio-snippet